File: 819. Most Common Word/819.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Solution {
public:
    string mostCommonWord(string paragraph, vector<string>& banned) {
        string str, result;
        int max_freq = 0;
        unordered_map<string, int> freq;
        unordered_set<string> banned_set(banned.begin(), banned.end());

        for(auto &c :paragraph)
            c = ispunct(c) ? ' ' : tolower(c);
        istringstream ss(paragraph);
        while(ss>>str)
            //freq[str]++;
            if(banned_set.count(str)==0 && max_freq<++freq[str]) {
                max_freq = freq[str];
                result = str;
            }
        return result;
    }
};

int main(){
    string paragraph = "Bob hit a ball, the hit BALL flew far after it was hit.";
    vector<string> banned = {"hit"};
    Solution sol;
    cout << sol.mostCommonWord(paragraph, banned) << endl;
    return 0;
}

// compile: g++ 819.cpp -o 819 -std=c++11
// execute: ./819

File: 819. Most Common Word/819.py
import string

class Solution(object):
    def mostCommonWord(self, paragraph, banned):
        """
        :type paragraph: str
        :type banned: List[str]
        :rtype: str
        """
        result = ""
        max_freq = 0
        freq = {}
        banned_set = set(banned)

        new_paragraph = ''.join([' ' if c in string.punctuation else c.lower() for c in paragraph])
        words = new_paragraph.split()
        for word in words:
            freq[word] = freq.get(word, 0) + 1
            # print(word, " ", freq[word])
            if (word not in banned_set) and (max_freq < freq[word]):
                max_freq = freq[word]
                result = word
        return result
        
if __name__ == "__main__":
    print(Solution().mostCommonWord("Bob hit a ball, the hit BALL flew far after it was hit.", ["hit"]))

# python3 819.py